relay-general/src/protocol/thread.rs

```diff
@@ -69,6 +71,114 @@ impl Empty for ThreadId {
+#[cfg_attr(feature = "jsonschema", derive(schemars::JsonSchema))]
+#[cfg_attr(feature = "jsonschema", schemars(rename_all = "lowercase"))]
+pub enum LockReasonType {
+    Locked = 1,
+    Waiting = 2,
+    Sleeping = 4,
+    Blocked = 8,
+}
+
+impl LockReasonType {
+    fn from_android_lock_reason_type(value: u64) -> Option<LockReasonType> {
+        Some(match value {
+            1 => LockReasonType::Locked,
+            2 => LockReasonType::Waiting,
+            4 => LockReasonType::Sleeping,
+            8 => LockReasonType::Blocked,
+            _ => return None,
+        })
+    }
+}
+
+impl fmt::Display for LockReasonType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            LockReasonType::Locked => write!(f, "locked"),
+            LockReasonType::Waiting => write!(f, "waiting"),
+            LockReasonType::Sleeping => write!(f, "sleeping"),
+            LockReasonType::Blocked => write!(f, "blocked"),
+        }
+    }
+}
+
+impl Empty for LockReasonType {
+    #[inline]
+    fn is_empty(&self) -> bool {
+        false
+    }
+}
+
+impl FromValue for LockReasonType {
+    fn from_value(value: Annotated<Value>) -> Annotated<Self> {
+        match value {
+            Annotated(Some(Value::U64(val)), mut meta) => {
+                match LockReasonType::from_android_lock_reason_type(val) {
+                    Some(value) => Annotated(Some(value), meta),
+                    None => {
+                        meta.add_error(ErrorKind::InvalidData);
+                        meta.set_original_value(Some(val));
+                        Annotated(None, meta)
+                    }
+                }
+            }
+            Annotated(Some(Value::I64(val)), mut meta) => {
+                match LockReasonType::from_android_lock_reason_type(val as u64) {
+                    Some(value) => Annotated(Some(value), meta),
+                    None => {
+                        meta.add_error(ErrorKind::InvalidData);
+                        meta.set_original_value(Some(val));
+                        Annotated(None, meta)
+                    }
+                }
+            }
+            Annotated(None, meta) => Annotated(None, meta),
+            Annotated(Some(value), mut meta) => {
+                meta.add_error(Error::expected("lock reason type"));
+                meta.set_original_value(Some(value));
+                Annotated(None, meta)
+            }
+        }
+    }
+}
+
+impl IntoValue for LockReasonType {
+    fn into_value(self) -> Value {
+        Value::String(self.to_string())
+    }
+
+    fn serialize_payload<S>(&self, s: S, _behavior: SkipSerialization) -> Result<S::Ok, S::Error>
+    where
+        Self: Sized,
+        S: Serializer,
+    {
+        Serialize::serialize(&self.to_string(), s)
+    }
+}
+
+impl ProcessValue for LockReasonType {}
+
+#[derive(Clone, Debug, PartialEq, Empty, FromValue, IntoValue, ProcessValue)]
```

Comment:
Please add a `///` doc comment above all public types and variants.

Reply:
ah yup, will do

Reply:
i'm going get roman to help me with docstrings for each of the LockReasonType enums (i'm not super familiar) but will get those in before merging! you can still review the other changes